[flang] Use module file hashes for more checking and disambiguation

klausler opened the pull request:
f18's module files are Fortran with a leading header comment containing the module file format version and a hash of the following contents.  This hash is currently used only to protect module files against corruption and truncation.

Extend the use of these hashes to catch or avoid some error cases.  When one module file depends upon another, note its hash in additional module file header comments.  This allows the compiler to detect when the module dependency is on a module file that has been updated.  Further, it allows the compiler to find the right module file dependency when the same module file name appears in multiple directories on the module search path.

The order in which module files are written, when multiple modules appear in a source file, is such that every dependency is written before the module(s) that depend upon it, so that their hashes are known.

A warning is emitted when a module file is not the first hit on the module file search path.

Further work is needed to add a compiler option that emits (larger) stand-alone module files that incorporate copies of their dependencies rather than relying on search paths. This will be desirable for application libraries that want to ship only "top-level" module files without needing to include their dependencies.

Another future work item would be to admit multiple modules in the same compilation with the same name if they have distinct hashes.

jeanPerier commented on flang/lib/Semantics/mod-file.cpp:
```diff
@@ -1250,6 +1333,47 @@ Scope *ModFileReader::Read(const SourceName &name,
+    }
+    if (!misses.empty()) {
+      auto &msg{Say(name, ancestorName,
+          "Could not find a module file for '%s' in the module search path with the expected checksum"_err_en_US,
```
It could be nice hinting to the user that some module may be outdated (and point to it, although maybe we do not know here which module file needs the module being opened.).

jeanPerier commented on flang/include/flang/Semantics/module-dependences.h:
```diff
@@ -0,0 +1,39 @@
+class ModuleDependences {
+public:
+  void AddDependence(std::string &&name, ModuleCheckSumType hash) {
+    map_.emplace(std::move(name), hash);
```
Maybe it would be worth complaining if there is already a conflicting hash that was added to prevent these bad usages (that may also be caused if File1 and File 2 were the same but File3 was outdated and not File4 or vice/versa).

klausler replied:
I think that I mentioned this in the commit comment.  Supporting multiple modules with the same name and distinct hashes is something we might consider doing in the future.  An attempt to do this now should get an error message about the module having the wrong checksum on its second use.

jeanPerier reviewed (approved):
I like the strategy.

Plus, as I understand the logic, the new "$!need" is optional (the check will just not be performed if it is not here), so this will not break existing .mod from people's build (they will just pick up the change the next time they rebuild the module with an updated flang).

klausler commented on flang/lib/Semantics/mod-file.cpp:
```diff
@@ -1155,12 +1182,17 @@ static bool FileContentsMatch(const std::string &path,
 // This uses the Fowler-Noll-Vo hash function.
-static std::string CheckSum(const std::string_view &contents) {
-  std::uint64_t hash{0xcbf29ce484222325ull};
+using ModuleCheckSumType = std::uint64_t;
```
No, this was just its original placement; removed.

klausler commented on flang/lib/Semantics/mod-file.cpp:
```diff
@@ -1169,18 +1201,69 @@ static std::string CheckSum(const std::string_view &contents) {
+    ModuleDependences &dependences, llvm::ArrayRef<char> content) {
+  std::size_t limit{content.size()};
+  std::string_view str{content.data(), limit};
+  for (std::size_t j{ModHeader::len}; str.substr(j, 7) == "!need$ ";) {
```
ok, done

klausler committed:
[flang] Use module file hashes for more checking and disambiguation

f18's module files are Fortran with a leading header comment
containing the module file format version and a hash of the
following contents.  This hash is currently used only to protect
module files against corruption and truncation.

Extend the use of these hashes to catch or avoid some error
cases.  When one module file depends upon another, note its
hash in additional module file header comments.  This allows
the compiler to detect when the module dependency is on a
module file that has been updated.  Further, it allows the
compiler to find the right module file dependency when the
same module file name appears in multiple directories on the
module search path.

The order in which module files are written, when multiple
modules appear in a source file, is such that every dependency
is written before the module(s) that depend upon it, so that
their hashes are known.

A warning is emitted when a module file is not the first hit
on the module file search path.

Further work is needed to add a compiler option that emits
(larger) stand-alone module files that incorporate copies of
their dependencies rather than relying on search paths.
This will be desirable for application libraries that want
to ship only "top-level" module files without needing to
include their dependencies.

Another future work item would be to admit multiple modules
in the same compilation with the same name if they have
distinct hashes.
